### PR TITLE
Refactoring of reranker training code (model, dataset) and bug fixes

### PR DESCRIPTION
**Pull request by Ziyang1060**
EN：
- Integrated SeqClassificationRanker for discriminative scoring training of BERT and LLM
- Integrated RankerDataset for binary classification, multi-label, distillation tasks
- Fixed the bug of learning rate warm-up in trainer.py code
- Support the yaml configuration file format for training, and categorize training parameters
- Supplement the README.md and add comprehensive parameter explanations

ZH：
- 整合 SeqClassificationRanker 用于 BERT 和 LLM 的判别式得分训练
- 整合 RankerDataset 用于二分类、多标签、蒸馏等任务
- 修复 trainer.py 代码中学习率预热的 bug
- 支持 yaml 配置文件格式用于训练，将训练参数分门别类
- 进一步补充README.md，增添全面的参数解释

### CODE REVIEW

File: rag_retrieval/train/reranker/README_zh.md [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+多卡训练配置文件修改:
+- 修改命令中的 CUDA_VISIBLE_DEVICES="0" 为你想要设置的多卡。
+- 修改上述提到的配置文件的 num_processes 为你想要跑的卡的数量。
+
+
+# 加载模型进行预测
+
+对于保存的模型，你可以很容易加载模型来进行预测。在modeling.py中，我们给出了一个例子。
+
+为了满足 LLM 如 "Qwen/Qwen2.5-1.5B" 用于判别式排序的特殊情况，设计了相关格式，实际效果为："query: {xxx} document: {xxx}\<score>"，实验显示 special_token 的引入对 LLM 排序性能提升较大。
```
[the remark this answers is not in the document]
Comment (Ziyang1060) [reply rag_retrieval/train/reranker/README_zh.md:129]: 该结论主要是从我自己的垂域实验结果得出的，并且也有相关论文提及：
Best Practices for Distilling Large Language Models into BERT for Web Search Ranking(https://arxiv.org/abs/2411.04539) ：
image

File: rag_retrieval/train/reranker/README_zh.md [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+对于 LLM 类模型，建议使用 deepspeed 来支持多卡训练模型，目前只支持 zero1 和 zero2 的训练阶段，以下是配置文件的示例。
+- [deepspeed_zero1](https://github.com/NLPJCL/RAG-Retrieval/blob/master/config/deepspeed/deepspeed_zero1.yaml)
+- [deepspeed_zero2](https://github.com/NLPJCL/RAG-Retrieval/blob/master/config/deepspeed/deepspeed_zero2.yaml)
+
+多卡训练配置文件修改:
+- 修改命令中的 CUDA_VISIBLE_DEVICES="0" 为你想要设置的多卡。
+- 修改上述提到的配置文件的 num_processes 为你想要跑的卡的数量。
+
+
+# 加载模型进行预测
```
[the remark this answers is not in the document]
Comment (Ziyang1060) [reply rag_retrieval/train/reranker/README_zh.md:125]: 好的，交由你处理～

File: rag_retrieval/train/reranker/trainer.py [hunk 3]
```diff
@@ -196,9 +232,11 @@ def on_epoch_end(self) -> None:
         self.progress_bar.close()
 
     def show_metrics(self, metrics: dict[str, float]) -> None:
-        description = f'Epoch {self.current_epoch}/{self.epochs}'
+        description = f"Epoch {self.current_epoch}/{self.epochs}"
         for name, score in metrics.items():
-            description += f' - {name}: {score:.6f}'
+            if "lr" in name:
```
[the remark this answers is not in the document]
Comment (Ziyang1060) [reply rag_retrieval/train/reranker/trainer.py:237]: 在分组前提下，可以加一个条件判定，取出 Group 0 或者 Group 1，如果目前全打印出来，进度条会很难看

File: rag_retrieval/train/reranker/modeling.py [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+
+        return all_logits
+
+    def preprocess(self, sentences_pairs, max_len):
+        temp = []
+        for query, document in sentences_pairs:
+            new_query = self.query_format.format(query.strip()) + self.seq
+            document_max_length = (
+                max_len
+                - len(self.tokenizer.encode(new_query))
```
[the remark this answers is not in the document]
Comment (Ziyang1060) [reply rag_retrieval/train/reranker/modeling.py:77]: 是的，需要加上 add_special_tokens=False 来保证计算 query 长度的准确性

File: rag_retrieval/train/reranker/modeling.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+                max_length=document_max_length,
+                truncation=True,
+                add_special_tokens=False,
+            )
+            new_document = self.tokenizer.decode(document_invalid_ids)
+            temp.append([new_query, new_document + self.special_token])
+        assert len(temp) == len(sentences_pairs)
+        sentences_pairs = temp
+
+        tokens = self.tokenizer.batch_encode_plus(
```
[the remark this answers is not in the document]
Comment (Ziyang1060) [reply rag_retrieval/train/reranker/modeling.py:91]: 仅仅依靠only_second是不行的，因为我们需要引入 special token，除非截断之后进行手动替换特殊 token，但这样其实就更加复杂了。当前这样做可以保证是对文档长度截断，对后续可能存在其他 format 格式也可以兼容。

**Comment by NLPJCL** [comment]
There are many modifications, so it is temporarily closed and waiting for a new PR to be submitted.
